### ISSUE TITLE
Separate file modification/access times (or, alternatively, dropping access times)

### ISSUE THREAD

**Issue by camilstaps** [emscripten-core/emscripten#11454]
From the documentation I understand that in the filesystem API currently one timestamp is kept per file, which serves as all of `atime`, `mtime`, and `ctime`. This is causing an issue in my application, which includes a kind of `make` tool that relies on modification times to check if files need to be regenerated.

- Are there plans for separating this single `timestamp`?
- Are there any fundamental problems with this?
- Is this something you want, and if so, would it be a good first issue (I have plenty of programming experience, but am not familiar with the code base)? Where to start?
- Or, alternatively, would it be feasible to add an option to disable updating of the timestamp on access only, which would not be really neat but would solve my current problem?

**Comment by camilstaps** [comment]
Looking at the code for `MEMFS` it seems quite feasible to add separate times, but I'm unsure about the others. With `IDBFS` there is the problem that it must be able to read something stored with the old version, but I don't think that's too difficult.

**Comment by kripken** [comment]
Interesting question. My concern with something like this is that it will increase code size for anything using the filesystem code. But maybe it's worth it for the extra functionality?

To avoid that issue, could this be done as a new FS, `MEMFSWITHTIMESTAMPS` (but a better name :wink: ) or some kind of optional plugin to `MEMFS`?

**Comment by sbc100** [comment]
If you are opting into POSIX filesystem emulation I think its reasonable to expect some overhead.   3 timestamps per file compared to 1 doesn't seem like a lot of data, and code size should be minimal, but the proof will be in the PR I suppose.

**Comment by camilstaps** [comment]
Thanks for your comments. I'd expect code size increase to be minimal, but there may be more data for some use cases. But implementing this as a separate FS first sounds like a good idea, makes it easy to compare. And if you decide it isn't worth putting it into master, I'd still be able to work with it. I have seen #9372, so I'll see if I can get something working like that!

**Comment by rth** [comment]
Another use case where not having separate modification/access time is not ideal is pyodide https://github.com/iodide-project/pyodide/issues/737. Python imports system seem to use a cache based on modification timestamps, and so newly created python files cannot be imported unless this cache is reset manually.

@camilstaps Just curious, did you experiment more with a separate FS approach since your last comment?

**Comment by camilstaps** [comment]
@rth I'm ashamed to say that I haven't, and it's not on the top of my list any more...

**Comment by rth** [comment]
No worries, thanks. It was just in case.

**Comment by nopid** [comment]
@camilstaps Trying to work around this issue for pyodide, having a proper parent update as in #13100 seems to fix most of our problems. Would having the timestamp behave more like a `mtime` be enough to fix the problems you encountered? Or would you need a more complete POSIX-like virtual filesystem?

**Comment by camilstaps** [comment]
@nopid thanks for checking back with me. It seems our issues are different, and mine has been resolved since posting the issue. I tried looking through the commit history but could not find where this might have happened. I may of course have been mistaken when posting the issue as well.

My issue, as I recall, was that file *reading* updated the timestamp. I have a make system where `file.y` is generated from `file.x`, and I use the timestamp to check whether I need to regenerate files. If I remember correctly, the timestamp of `file.x` was getting updated even if I only accessed the file for reading, causing `file.y` to be regenerated all the time. However, this issue does not seem to be present any more.

So my issue was unrelated to directory timestamps, but looking at #13100 I think the changes make sense.

@sbc100 I think this issue can be closed.
